src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -0,0 +1,234 @@
+                WebUtility.UrlEncode("c"),
+                WebUtility.UrlEncode(sig),
+                WebUtility.UrlEncode(signedstart),
+                WebUtility.UrlEncode(signedexpiry),
+                WebUtility.UrlEncode(signedpermissions)
+                );
+
+            return sasURL;
+        }
+
+
+        public static string AuthorizationHeader(string StorageAccount, string StorageKey, string method, DateTime now,
+            HttpWebRequest request, string ifMatch = "", string md5 = "", string size="",string contentType="")
+        {
+            var MessageSignature = string.Format("{0}\n\n\n{1}\n{5}\n{6}\n\n\n{2}\n\n\n\n{3}{4}",
+                method,
+                (size == String.Empty) ? string.Empty : size,
+                ifMatch,
+                GetCanonicalizedHeaders(request),
+                GetCanonicalizedResource(request.RequestUri, StorageAccount),
+                md5,
+                contentType
+                );
+
+            var SignatureBytes = Encoding.UTF8.GetBytes(MessageSignature);
+            var SHA256 = new HMACSHA256(Convert.FromBase64String(StorageKey));
+            var AuthorizationHeader = "SharedKey " + StorageAccount + ":" +
+                                      Convert.ToBase64String(SHA256.ComputeHash(SignatureBytes));
+            return AuthorizationHeader;
+        }
+
+        public static string GetCanonicalizedHeaders(HttpWebRequest request)
+        {
+            var headerNameList = new ArrayList();
+            var sb = new StringBuilder();
+            foreach (var headerName in request.Headers.AllKeys)
+            {
+                if (headerName.ToLowerInvariant().StartsWith("x-ms-", StringComparison.Ordinal))
+                {
+                    headerNameList.Add(headerName.ToLowerInvariant());
+                }
+            }
+            headerNameList.Sort();
+            foreach (string headerName in headerNameList)
+            {
+                var builder = new StringBuilder(headerName);
+                var separator = ":";
+                foreach (string headerValue in GetHeaderValues(request.Headers, headerName))
+                {
+                    var trimmedValue = headerValue.Replace("\r\n", string.Empty);
+                    builder.Append(separator);
+                    builder.Append(trimmedValue);
+                    separator = ",";
+                }
+                sb.Append(builder);
+                sb.Append("\n");
+            }
+            return sb.ToString();
+        }
+
+        public static ArrayList GetHeaderValues(WebHeaderCollection headers, string headerName)
+        {
+            var list = new ArrayList();
+            var values = headers[headerName];
+            if (values != null)
+            {
+                list.Add(values.TrimStart(null));
+            }
+            return list;
+        }
+
+        public static string GetCanonicalizedResource(Uri address, string accountName)
```

Comment:
This method is 100 lines long. It might be a bit easier to digest if it the giant foreach loops were factored into private methods. I can't remember what the types of all the local variables are with the method so large.